src/wasm/wat-parser.cpp

```diff
@@ -0,0 +1,1809 @@
+
+// start ::= '(' 'start' x:funcidx ')'
+
+// elem ::= '(' 'elem' id? (et,y*):elemlist ')'
+//        | '(' 'elem' id? x:tableuse '(' 'offset'? e:expr ')'
+//              (et,y*):elemlist ')'
+//        | '(' 'elem' id? 'declare' (et,y*):elemlist ')'
+// elemlist ::= t:reftype (y:elemexpr)*
+// elemexpr ::= '(' 'item' e:expr ')'
+//            | '(' e:expr ')'
+// tableuse ::= '(' 'table' x:tableidx ')'
+
+// data ::= '(' 'data' id? b:datastring ')'
+//        | '(' 'data' id? x:memuse '(' 'offset'? expr ')' b:datastring ')'
+// memuse ::= ('(' 'memory' x:memidx ')')?
+
+// modulefield ::= type
+//               | import
+//               | func
+//               | table
+//               | mem
+//               | global
+//               | export
+//               | start
+//               | elem
+//               | data
+Result<> modulefield(ParseDeclsCtx& ctx, ParseInput& in) {
+  if (type(ctx, in)) {
+    return Ok{};
+  }
+  if (import(ctx, in)) {
+    return Ok{};
+  }
+  if (func(ctx, in)) {
+    return Ok{};
+  }
+  if (table(ctx, in)) {
+    return Ok{};
+  }
+  if (mem(ctx, in)) {
+    return Ok{};
+  }
+  if (global(ctx, in)) {
+    return Ok{};
+  }
+  // if (export(ctx, in)) {
+  //   return Ok{};
+  // }
+  // if (start(ctx, in)) {
+  //   return Ok{};
+  // }
+  // if (elem(ctx, in)) {
+  //   return Ok{};
+  // }
+  // if (data(ctx, in)) {
+  //   return Ok{};
+  // }
+  return {};
+}
+
+// module ::= '(' 'module' id? (m:modulefield)* ')'
+//          | (m:modulefield)* eof
+Result<> module(ParseDeclsCtx& ctx, ParseInput& in) {
+  bool outer = in.takeSExprStart("module"sv);
+
+  if (outer) {
+    if (auto id = in.takeID()) {
+      ctx.wasm.name = *id;
+    }
+  }
+
+  while (modulefield(ctx, in)) {
+  }
+
+  if (outer && !in.takeRParen()) {
+    return {};
+  }
+
+  return Ok{};
+}
+
+void applyImportNames(Importable& item,
+                      const std::optional<ImportNames>& names) {
+  if (names) {
+    item.module = names->mod;
+    item.base = names->nm;
+  }
+}
+
+Result<> addExports(Module& wasm,
+                    const Named* item,
+                    const std::vector<Name>& exports,
+                    ExternalKind kind) {
+  for (auto name : exports) {
+    if (wasm.getExportOrNull(name)) {
+      return {};
+    }
+    wasm.addExport(Builder(wasm).makeExport(name, item->name, kind));
+  }
+  return Ok{};
+}
+
+Result<IndexMap> createIndexMap(const std::vector<DefSpan>& defs) {
+  IndexMap indices;
+  for (Index i = 0; i < defs.size(); ++i) {
+    if (defs[i].name.is()) {
+      bool inserted = indices.insert({defs[i].name, i}).second;
+      if (!inserted) {
+        return {};
+      }
+    }
+  }
+  return indices;
+}
+
+} // anonymous namespace
+
+Result<> parseModule(Module& wasm, std::string_view input) {
+  // Parse module-level declarations and implicit type definitions.
+  ParseDeclsCtx decls(wasm);
+  {
+    ParseInput in(input);
+    if (!module(decls, in)) {
+      return {};
+    }
```

Comment:
How will error handling work? For example here, shouldn't we somehow return the source location on which we failed to find a module? Or should the caller somehow handle that?

Reply:
I'm in the process of updating `Result<T>` to be a variant that either holds a `T` or holds an `Err` type that contains a message and a location. Everywhere we currently have `return {};`, we will either propagate an error or generate a new one instead.

Reply:
Just pushed an update with error handling--PTAL! I'm next going to introduce a `MaybeResult<T>` as a more ergonomic return value for all the parser functions that are currently returning either bool or `std::optional<T>` depending on the parsing stage.